File: app/src/main/java/de/christinecoenen/code/zapp/views/ProgramInfoViewBase.java
package de.christinecoenen.code.zapp.views;


import android.content.Context;
import android.databinding.ViewDataBinding;
import android.os.Handler;
import android.util.AttributeSet;
import android.util.Log;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.widget.LinearLayout;

import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.TimeUnit;

import butterknife.BindInt;
import butterknife.ButterKnife;
import de.christinecoenen.code.programguide.ProgramGuideRequest;
import de.christinecoenen.code.programguide.model.Show;
import de.christinecoenen.code.zapp.R;
import de.christinecoenen.code.zapp.model.ChannelModel;

public abstract class ProgramInfoViewBase extends LinearLayout {

	private static final String TAG = ProgramInfoViewBase.class.getSimpleName();


	@BindInt(R.integer.view_program_info_update_show_info_interval_seconds)
	protected int updateShowInfoIntervalSeconds;

	@BindInt(R.integer.view_program_info_update_show_time_interval_seconds)
	protected int updateShowTimeIntervalSeconds;

	private ProgramGuideRequest currentShowInfoRequest;
	private Show currentShow = null;
	private ChannelModel currentChannel;
	private ViewDataBinding binding;

	private final Handler handler = new Handler();
	private Timer timer;

	private final ProgramGuideRequest.Listener programGuideListener = new ProgramGuideRequest.Listener() {
		@Override
		public void onRequestError() {
			logMessage("could not load show info");

			ProgramInfoViewBase.this.currentShow = null;
			applyShow(currentShow);
			setIsLoading(false);
		}

		@Override
		public void onRequestSuccess(Show currentShow) {
			logMessage("show info loaded: " + currentShow);

			ProgramInfoViewBase.this.currentShow = currentShow;
			applyShow(currentShow);
			setIsLoading(false);
		}
	};

	public ProgramInfoViewBase(Context context, AttributeSet attrs) {
		super(context, attrs);

		LayoutInflater inflater = (LayoutInflater) context
			.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
		binding = inflate(inflater);

		setOrientation(LinearLayout.VERTICAL);
		setGravity(Gravity.CENTER_VERTICAL);

		ButterKnife.bind(this, this);
	}

	public ProgramInfoViewBase(Context context) {
		this(context, null);
	}

	public void setChannel(ChannelModel channel) {
		if (channel == currentChannel) {
			return;
		}

		currentShow = null;
		currentChannel = channel;

		applyShow(null);

		chancelProgramGuideLoading();
		loadProgramGuide();
	}

	public void pause() {
		timer.cancel();
		timer = null;
	}

	public void resume() {
		if (timer != null) {
			return;
		}

		timer = new Timer();
		timer.scheduleAtFixedRate(new UpdateShowInfoTask(), 0,
			TimeUnit.SECONDS.toMillis(updateShowInfoIntervalSeconds));
		timer.scheduleAtFixedRate(new UpdateShowTimeTask(), 0,
			TimeUnit.SECONDS.toMillis(updateShowTimeIntervalSeconds));
	}

	protected abstract ViewDataBinding inflate(LayoutInflater inflater);

	protected abstract void applyShow(Show show);

	protected abstract void setIsLoading(boolean isLoading);

	private void updateShowInfo() {
		if (currentShow == null ||
			currentShow.getEndTime() == null ||
			currentShow.getEndTime().isBeforeNow()) {
			reloadProgramGuide();
		}
	}

	private void displayTime() {
		binding.invalidateAll();
	}

	private void reloadProgramGuide() {
		if (currentChannel == null) {
			return;
		}

		logMessage("reloadProgramGuide");
		chancelProgramGuideLoading();
		loadProgramGuide();
	}

	private void loadProgramGuide() {
		setIsLoading(true);

		currentShowInfoRequest = new ProgramGuideRequest(getContext())
			.setChannelId(currentChannel.getId())
			.setListener(programGuideListener)
			.execute();
	}

	private void chancelProgramGuideLoading() {
		setIsLoading(false);
		if (currentShowInfoRequest != null) {
			currentShowInfoRequest.cancel();
		}
	}

	private void logMessage(String message) {
		String channelId = null;
		if (currentChannel != null) {
			channelId = currentChannel.getId();
		}
		Log.d(TAG, channelId + " - " + message);
	}

	private class UpdateShowTimeTask extends TimerTask {
		@Override
		public void run() {
			handler.post(new Runnable() {
				@Override
				public void run() {
					displayTime();
				}
			});
		}
	}

	private class UpdateShowInfoTask extends TimerTask {
		@Override
		public void run() {
			handler.post(new Runnable() {
				@Override
				public void run() {
					logMessage("UpdateShowInfoTask");
					updateShowInfo();
				}
			});
		}
	}
}

File: app/src/main/java/de/christinecoenen/code/zapp/app/livestream/ui/views/ProgramInfoViewBase.kt
package de.christinecoenen.code.zapp.app.livestream.ui.views

import android.animation.AnimatorInflater
import android.animation.ObjectAnimator
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.text.format.DateUtils
import android.util.AttributeSet
import android.view.Gravity
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import androidx.core.view.isVisible
import de.christinecoenen.code.zapp.R
import de.christinecoenen.code.zapp.app.livestream.model.LiveShow
import de.christinecoenen.code.zapp.app.livestream.repository.ChannelInfoRepository
import de.christinecoenen.code.zapp.models.channels.ChannelModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import timber.log.Timber
import java.util.*
import java.util.concurrent.TimeUnit
import kotlin.math.roundToInt

abstract class ProgramInfoViewBase @JvmOverloads constructor(
	context: Context,
	attrs: AttributeSet? = null
) : LinearLayout(context, attrs) {

	private val uiHandler = Handler(Looper.getMainLooper())

	private val updateShowInfoIntervalSeconds =
		resources.getInteger(R.integer.view_program_info_update_show_info_interval_seconds)

	private val updateShowTimeIntervalSeconds =
		resources.getInteger(R.integer.view_program_info_update_show_time_interval_seconds)

	private var currentShow: LiveShow? = null
	private var currentChannel: ChannelModel? = null
	private var timer: Timer? = null
	private var showJob: Job? = null

	private val showTitleView: TextView by lazy {
		rootView.findViewById(R.id.text_show_title)
	}
	private val showSubtitleView: TextView by lazy {
		rootView.findViewById(R.id.text_show_subtitle)
	}
	private val showTimeView: TextView by lazy {
		rootView.findViewById(R.id.text_show_time)
	}
	private val progressBarView: ProgressBar by lazy {
		rootView.findViewById(R.id.progressbar_show_progress)
	}

	private val showProgressAnimator: ObjectAnimator by lazy {
		val animator = AnimatorInflater.loadAnimator(context, R.animator.view_program_info_show_progress) as ObjectAnimator
		animator.target = progressBarView
		animator
	}


	init {
		orientation = VERTICAL
		gravity = Gravity.CENTER_VERTICAL
	}


	fun setChannel(channel: ChannelModel) {
		if (channel === currentChannel) {
			return
		}

		currentShow = null
		currentChannel = channel

		showTitleView.text = ""
		showSubtitleView.text = ""
		showTimeView.text = ""
		progressBarView.progress = 0

		chancelProgramGuideLoading()
		loadProgramGuide()
	}

	fun pause() {
		timer?.cancel()
		timer = null
	}

	fun resume() {
		if (timer != null) {
			return
		}

		timer = Timer().apply {
			scheduleAtFixedRate(UpdateShowInfoTask(), 0,
				TimeUnit.SECONDS.toMillis(updateShowInfoIntervalSeconds.toLong()))
			scheduleAtFixedRate(UpdateShowTimeTask(), 0,
				TimeUnit.SECONDS.toMillis(updateShowTimeIntervalSeconds.toLong()))
		}
	}

	private fun onRequestError(e: Throwable) {
		logMessage("could not load show info: " + e.message)

		currentShow = null

		showTitleView.setText(R.string.activity_channel_detail_info_error)

		showSubtitleView.isVisible = false
		showTimeView.isVisible = false
		progressBarView.isVisible = false
	}

	private fun onRequestSuccess(newShow: LiveShow) {
		logMessage("show info loaded: ${newShow.title}")

		currentShow = newShow

		displayTitles()
		displayTime()
	}

	private fun updateShowInfo() {
		if (currentShow?.endTime?.isBeforeNow == true) {
			reloadProgramGuide()
		}
	}

	private fun displayTime() {
		if (currentShow == null) {
			return
		}

		if (currentShow!!.hasDuration()) {
			val progressPercent = currentShow!!.progressPercent
			val progress = (progressPercent * progressBarView.max).roundToInt()

			val startTime = DateUtils.formatDateTime(context,
				currentShow!!.startTime!!.millis, DateUtils.FORMAT_SHOW_TIME)
			val endTime = DateUtils.formatDateTime(context,
				currentShow!!.endTime!!.millis, DateUtils.FORMAT_SHOW_TIME)
			val fullTime = context.getString(R.string.view_program_info_show_time, startTime, endTime)

			showTimeView.text = fullTime
			showTimeView.isVisible = true
			progressBarView.isIndeterminate = false
			progressBarView.isEnabled = true
			setShowProgressBar(progress)
		} else {
			setShowProgressBar(0)
			showTimeView.isVisible = false
			progressBarView.isIndeterminate = false
			progressBarView.isEnabled = false
		}

		progressBarView.isVisible = true
	}

	private fun displayTitles() {
		if (currentShow == null) {
			return
		}

		showTitleView.text = currentShow!!.title

		if (currentShow!!.hasSubtitle()) {
			showSubtitleView.text = currentShow!!.subtitle
			showSubtitleView.isVisible = true
		} else {
			showSubtitleView.isVisible = false
		}
	}

	private fun reloadProgramGuide() {
		if (currentChannel == null) {
			return
		}

		logMessage("reloadProgramGuide")
		chancelProgramGuideLoading()

		loadProgramGuide()
	}

	private fun loadProgramGuide() {
		progressBarView.isEnabled = true
		progressBarView.isIndeterminate = true

		showJob = GlobalScope.launch(Dispatchers.Main) {
			try {
				val show = ChannelInfoRepository.getInstance().getShows(currentChannel!!.id)
				onRequestSuccess(show)
			} catch (e: Exception) {
				onRequestError(e)
			}
		}
	}

	private fun chancelProgramGuideLoading() {
		progressBarView.isIndeterminate = false
		showJob?.cancel()
		showJob = null
	}

	private fun setShowProgressBar(value: Int) {
		showProgressAnimator.setIntValues(value)
		showProgressAnimator.start()
	}

	private fun logMessage(message: String) {
		Timber.d("${currentChannel?.id} - $message")
	}

	private inner class UpdateShowTimeTask : TimerTask() {
		override fun run() {
			uiHandler.post { displayTime() }
		}
	}

	private inner class UpdateShowInfoTask : TimerTask() {
		override fun run() {
			uiHandler.post {
				logMessage("UpdateShowInfoTask")
				updateShowInfo()
			}
		}
	}
}
